client/state/posts/utils.js

```diff
@@ -91,3 +91,45 @@ export function mergeIgnoringArrays( object, ...sources ) {
 		}
 	} );
 }
+
+/**
+ * Takes existing term post edits and updates the `terms_by_id` attribute
+ *
+ * @param  {Object}    post  object of post edits
+ * @return {Object}          normalized post edits
+ */
+export function getTermIdsFromEdits( post ) {
+	if ( ! post || ! post.terms ) {
+		return post;
+	}
+
+	// Skip "default" taxonomies until legacy token-field and category-selector are removed
+	let taxonomies = omit( post.terms, [ 'post_tag', 'category' ] );
+
+	// Filter taxonomies that are set as arrays ( i.e. tags )
+	// This can be detected by an array of strings vs an array of objects
+	taxonomies = reduce( taxonomies, ( prev, taxonomyTerms, taxonomyName ) => {
+		// Ensures we are working with an array
+		const termsArray = map( taxonomyTerms );
+		if ( termsArray && termsArray.length && ! isObject( termsArray[ 0 ] ) ) {
+			return prev;
+		}
+
+		prev[ taxonomyName ] = termsArray;
+		return prev;
+	}, {} );
+
+	if ( isEmpty( taxonomies ) ) {
+		return post;
+	}
+
+	post.terms_by_id = mapValues( taxonomies, ( taxonomy ) => {
```

Comment:
I believe this is mutating the original post object.

In `getNormalizedPost`, we use `cloneDeep` to create an object clone, but that's mostly an artifact of `lib/post-normalizer` mutating the original objects. I'd say we should avoid this for these new normalizations and try to not mutate the object in the first place.